src/system/SystemPacketBuffer.h

```diff
@@ -230,3 +243,91 @@ inline uint16_t PacketBuffer::AllocSize() const
 
 } // namespace System
 } // namespace chip
+
+namespace chip {
+namespace System {
+
+/// Tracks ownership of a System::PacketBuffer.
+class DLL_EXPORT PacketBufferHandle
+{
+public:
+    PacketBufferHandle() : mBuffer(nullptr) {}
+    PacketBufferHandle(decltype(nullptr)) : mBuffer(nullptr) {}
+
+    PacketBufferHandle(PacketBufferHandle && aOther)
+    {
+        mBuffer        = aOther.mBuffer;
+        aOther.mBuffer = nullptr;
+    }
+
+    ~PacketBufferHandle() { Adopt(nullptr); }
+
+    PacketBufferHandle & operator=(PacketBufferHandle && aOther)
+    {
+        if (mBuffer != nullptr)
+        {
+            PacketBuffer::Free(mBuffer);
+        }
+        mBuffer        = aOther.mBuffer;
+        aOther.mBuffer = nullptr;
+        return *this;
+    }
+    PacketBufferHandle & operator=(decltype(nullptr))
+    {
+        Adopt(nullptr);
+        return *this;
+    }
+
+    PacketBufferHandle Clone() const
```

Comment:
Requesting opinions for a future PR — I now dislike this name. Typical use is `DoSomething(buffer.Clone())` which gives the impression to the unfamiliar reader that the buffer (rather than the handle) is being cloned.

Reply:
My personal preference would be rename this `Retain` since it has the underlying effect of increasing the reference count of the underlying buffer by one (which is the important operation). That a handle is being clone is somewhat immaterial just as in a std::shared_ptr passed by value creates a clone of the shared_ptr. It's what the shared_ptr points to that is important and that a strong reference to it is being retained.

Reply:
Yeah, `Retain` might work.  Another possibility might be `Share`, since we are now sharing ownership with the new handle?

Reply:
I tend to prefer paradigms and names developers are familiar with. I'm a CoreFoundation / Cocoa (pre-ARC) guy, so `*Retain`/`*Release` (for example, `CFRetain`/`CFRelease`) naming and semantics are both familiar and meaningful to me.

Reply:
I like `Retain`. I'll switch to that in the next PR if there are no objections.

Reply:
I would find `Retain` readable/understandable as well. 
As another option just in case, we also have AddRef used in various places. No 'DecRef' though so Retain/Release pairing seems a bit more natural.